xrtl/gfx/pixel_format.h

```diff
@@ -0,0 +1,577 @@
+  kEac = 12 | kCompressedMask,   // 4x4 block = 16 bytes
+
+  // ASTC block-compressed format.
+  // https://en.wikipedia.org/wiki/Adaptive_Scalable_Texture_Compression
+  // https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#appendix-compressedtex-astc
+  kAstc = 13 | kCompressedMask,
+
+  // PVRTC block-compressed format as used by PowerVR chipsets, commonly iOS.
+  // https://en.wikipedia.org/wiki/PVRTC
+  kPvrtc = 14 | kCompressedMask,
+};
+
+// Component data type used in pixel formats.
+// This defines how the values are sampled in shaders. The widths and range of
+// each type is based on the component widths
+enum class ComponentFormat : uint8_t {
+  // Signed floating-point value.
+  // 16-bit floating-point formats use half-precision (s10e5 format).
+  // 32-bit floating-point formats use IEEE-754 single-precision (s23e8 format).
+  //
+  // Suffixes:
+  //  D3D12: _FLOAT
+  //  Metal: Float
+  //  Vulkan: _SFLOAT
+  kSFloat = 0,
+
+  // Unsigned floating-point value, used by some packed formats.
+  //
+  // Suffixes:
+  //  D3D12: _FLOAT
+  //  Metal: Float
+  //  Vulkan: _UFLOAT
+  kUFloat = 1,
+
+  // Signed normalized integer of some number of bits.
+  // When sampled in shaders this is normalzed to [-1.0, 1.0].
+  //
+  // Example:
+  //  kSNorm with size 2 as [-2, -1, 0, 1] is sampled as [-1.0, ..., 0.0, 1.0].
+  //
+  // Suffixes:
+  //  D3D12: _SNORM
+  //  Metal: Snorm
+  //  Vulkan: _SNORM
+  kSNorm = 2,
+
+  // Unsigned normalized integer of some number of bits.
+  // When sampled in shaders this is normalzed to [0.0, 1.0].
+  //
+  // Example:
+  //  kUNorm with size 2 as [0, 1, 2, 3] is sampled as [0.0, 1/3.0, 2/3.0, 1.0].
+  //
+  // Suffixes:
+  //  D3D12: _UNORM
+  //  Metal: Unorm
+  //  Vulkan: _UNORM
+  kUNorm = 3,
+
+  // Signed integer of some number of bits.
+  // This is sampled in shaders with no normalization.
+  //
+  // Example:
+  //  kSInt with size 2 is [-2, -1, 0, 1].
+  //
+  // Suffixes:
+  //  D3D12: _SINT
+  //  Metal: Sint
+  //  Vulkan: _SINT
+  kSInt = 4,
+
+  // Unsigned integer of some number of bits.
+  // This is sampled in shaders with no normalization.
+  //
+  // Example:
+  //  kUInt with size 2 is [0, 1, 2, 3].
+  //
+  // Suffixes:
+  //  D3D12: _UINT
+  //  Metal: Uint
+  //  Vulkan: _UINT
+  kUInt = 5,
+
+  // Signed floating-point value that is mapped through the sRGB ramp.
+  // Values of 0.0 will sample as 0.0 and 1.0 as 1.0, but the values inbetween
+  // will be modified. Alpha channels, if present, will never be adjusted.
+  //
+  // Suffixes:
+  //  D3D12: _SRGB
+  //  Metal: _sRGB/_srgb
+  //  Vulkan: _SRGB
+  kSrgb = 6,
+};
+
+// Each PixelFormat is defined by a unique ID combined with many packed flags.
+// This allows us to encode the most commonly used information directly into the
+// enum value and avoid indirections during data lookups. The unique ID allows
+// us to use tables to map to internal formats.
+class PixelFormat {
```

Comment:
Correct - images and such come a bit later - these just define how pixels are stored. This is required to start talking about things like what surface formats are used when creating OS windows/contexts/etc. They'll then be reused for when we talk about images/textures/samplers/etc.